### ISSUE TITLE
custom dataset creation for unisar

### ISSUE THREAD

**Issue by srewai** [microsoft/ContextualSP#37]
Hello @SivilTaram, 

I've two tables and they can be linked with primary and foreign key. I would like to use unisar on these tables. could you please share steps or hints on how can i create custom dataset for making use of unisar? I appreciate your help:)

**Comment by SivilTaram** [comment]
Maybe you can try to convert your dataset into the format of Spider as the first step? Then you may directly use UniSAR on these datasets! Ping @DreamerDeo for further help if needed.

**Comment by DreamerDeo** [comment]
Hi @srewai , many thanks for your interest in unisar.

To make the unisar work on your own dataset, the ONLY job is to call the step2 to generate the file following the same format as [preprocessed file like spider](https://github.com/microsoft/ContextualSP/tree/master/unified_parser_text_to_sql/dataset_post/spider_sl). If you don't want to involve the schema-linking information (e.g. exact-match, partial match and value match). You could just skip the step1.py. The performance difference is about 3% in Spider w/wo schema-linking.

You could simply transform the data format of yours into Spider-fashion as @SivilTaram mentioned (the easiest way). Moreover, other Spider codebase (e.g., IRNET, RATSQL) could be employed too. 
Otherwise, here is the place I think your need to take care of in unisar code:
1. the most tricky part is to derive the schema-linking tag(step1.py). The different format between yours and Spider will make some function(e.g., DBContext) in step1 failed. Luckliy, it's not  `derive_schema_linking_function` by yourself. The input of function are question, table headers and table values. The output should be the schema linking information following https://github.com/microsoft/ContextualSP/blob/2f9777ee68a2eed32c327c6c64d30d2e0420fce7/unified_parser_text_to_sql/semparse/contexts/spider_db_context.py#L125. 
3.  Make sure the schema-linking file follows the same format as [spider_after_schema_linking].(https://github.com/microsoft/ContextualSP/blob/master/unified_parser_text_to_sql/data/spider_schema_linking_tag/dev.json) Then you could directly call step2.py

Note that the current version of unisar is built on BART-large (English). That is, if the non-English parser is desired, you need to use the mbart pre-process (step2.py) and train command (train.py) as stated in fairseq documents.

If your need further help, please let me know. I'm very glad about that :)

**Comment by srewai** [comment]
first of all thank you very much @DreamerDeo and @SivilTaram. 
@DreamerDeo , amazing that you shared the detailed steps to data preparation. I am going to implement the same and get back in case of any concerns. It's great that you went extra mile and mentioned non-eng parser.
In addition to eng parser, i would like to test mbart as well. Do you think mbart is good choice for parsing German tables? 

In case , thank you so much once again! I will be back soon with some questions :D

**Comment by DreamerDeo** [comment]
@srewai  Actually, mBART shows a promising (but not as English) performance in German (i.e., question and DB are in German and the output SQL also involves German headers.) 

In our other project about multilingual text-to-SQL, we compare the (1)mBART-CC25 with (2)RAT-SQL + XLM-R-Large (This dataset/paper will be released later this month. I will pin you on this issue. Welcome to follow:) ). 
The set-match accuracy of mBART(without constrained decoding) is about 40% while RAT-SQL achieves 62%.
Through the case studies (this is done by native-speaker vendors). The most bad cases of vanilla mBART could be categorized to the synonym problem (e.g., nation -> citizen). This problem could be alleviated by careful design the way of construction of prefix-tree (it's a bit tricky but not very hard.)

Of course, you could try mBART first since it's more scalable and training is extremely fast (thanks to fairseq).
But if you have enough computational resources, I would recommend attempting mT5 rather than mBART in non-English. It's more powerful (a lot of multilingual NLP work also prefer mT5).

**Comment by srewai** [comment]
@DreamerDeo , that would be amazing. Please pin me on the release of the multilingual paper. I'm waiting for it:)
For now I only have 2 tables without "NL questions" and therefore, i don't think it is possible to create  the dataset. However, i would like to perform unisar inference on the these tables. I have processed the cleaned the tables. So if i understand correctly to perform inference on unisar ,  I will need to create sqlite db out of these table correct? Or rather would please share the steps for inference on new tables?
Thank you.

**Comment by DreamerDeo** [comment]
@srewai 
You can wait for our multilingual-Spider to train your own parser. It supports Chinese/Japanese/German/French/Spanish/Vietnamese. About 10K data and 120 tables.

Q1:  I will need to create sqlite db out of these table correct?
No. Our code could read the databse values from .json too.  Try this function https://github.com/microsoft/ContextualSP/blob/2f9777ee68a2eed32c327c6c64d30d2e0420fce7/unified_parser_text_to_sql/semparse/sql/spider_utils.py#L324

This function would return the values follows the same format like 
https://github.com/microsoft/ContextualSP/blob/2f9777ee68a2eed32c327c6c64d30d2e0420fce7/unified_parser_text_to_sql/semparse/sql/spider_utils.py#L111

Q2: the steps for inference on new tables
If you  want to **offline** predict, just serial your table schema and question into the format like https://github.com/microsoft/ContextualSP/tree/master/unified_parser_text_to_sql/dataset_post/spider_sl

If you want **online** inference like a demo, please try
https://github.com/microsoft/ContextualSP/blob/2f9777ee68a2eed32c327c6c64d30d2e0420fce7/unified_parser_text_to_sql/unisar/api.py#L55
You could just enter your question about the table, the commandline would return the predicted SQL.

**Comment by srewai** [comment]
@DreamerDeo , great! Let me try that out. So first i will need to convert my excel to json post which I can use the function:

`def read_dataset_values_from_json(db_id: str, db_content_dict: Dict[str, Any], tables: List[str]): `

**Comment by longxudou** [comment]
Hi @srewai, this is the paper about multilingualSpider: https://arxiv.org/pdf/2212.13492.pdf . Welcome to read.

The codebase and dataset will be released in about one month. (I need to prepare more time for ACL deadline )
At the dataset release time, I will pin you as well.

Since this issue is not activated for other person. I need to close it.

**Comment by epejhan90** [comment]
Hi,
Thank you for the explanation. But I still have three more questions regarding to run the model for my own data. I could run the three steps of unified_parser_text_to_sql for spider data and also run the interactive options. I would like to be able to query on my own data.
1. I have one table at this moment. To be able to add my data to the db , I have made the schema for it and also added the table to the tables.json. The other thing I should do manually is to add some queries about my table to the dev.son? 
My general question is what are the preprocesses to add my table to db and be able to run three steps?

2. My table also have timestamp columns, does the model also support that?
3. How many samples would be fine to add for my new db?

Thank you.

**Comment by SivilTaram** [comment]
@longxudou Maybe longxu can help on this question.

**Comment by longxudou** [comment]
@epejhan90  Thanks for your interest!

Q1: What are the preprocesses to add my table to db and be able to run three steps?

The codebase doesn't support this function. But you can implement it following your statement. I think it's reasonable to append the schema of new db in tables.json. But you should also append content of db in another via (1) creating db sqlite file like Spider does; or (2) storing all the content in another json file.

Note that the current version of UNISAR codebase only support non-value SQL generation. This is because of (1)the research benchmark mainly focus on SQL-sketch accuracy rather than SQL execution accuracy; (2) shorten the input length to avoid the length exceeding problem.

If you want to make your SQL executable, just append the value after corresponding columns. It works in my experiments.

Q2: My table also have timestamp columns, does the model also support that?

Firstly, I don't take into account the timestamp cases. But I think you can just treat "timestamp" as a special column. You don't have to change the pipeline except replacing this column with the concrete values after SQL generation.

Q3: How many samples would be fine to add for my new db?

If the domains of your db (assumed in English) are overlapped with Spider Train-set, I think it's unnecessary to further fine-tune the model.
See https://github.com/ryanzhumich/editsql/blob/master/data/spider/train_db_ids.txt for the details of domain information in Spider.

If you want to fine-tune the model for better performance, I think fifty cases is enough and also scalable.

**Comment by epejhan90** [comment]
@longxudou Thank you so much for your reply . I could fix it.

**Comment by longxudou** [comment]
@epejhan90  This is because the released checkpoint is no-value version, which is trained by "query_toks_no_value" rather than "query_toks" in Spider dataset. If you want to fix

**Comment by epejhan90** [comment]
@longxudou Thank you. I don't have access to any GPU at this moment and I'm running the code on google Colab. Do you think is it possible to retrain the model there?

**Comment by longxudou** [comment]
@epejhan90  I think you could fisrt fine-tune the non-value to be "valuable". Otherwise, it's time-consuming to adopt one GPU to fine-tune the BART-large model. In my experiments I adopted four V100-16G gpus for 10 hours. If you are not in hurry, you could wait for our released model in next month.
